EVG-6718 use $merge for hourly and daily test stats.

jimoleary opened the pull request:
Some things to note:

1. The new implementation follows the old, this allows them to be somewhat interchangeable and should throttle $merge.
1. It is possible to make the $merge restartable (although a *$sort* and *$limit* stage in the $merge pipelines would be required). The current implementation assumes that the job will run to completion. Both approaches likely require a new index in testresults (I'm pretty sure we can use the same index). 
1. The $merge work flow should now be idempotent. I'm pretty sure the existing one is not.
1. The $merge hourly and daily test stats no longer depend on each other. This is mainly for simplicity, but it also means that they could be run independantly (off cyce with each other). It does mean that we process the same testresults documents twice.
1. The $merge essentially works as if oldTasks is always enabled. It is not possible to disable it currently, fixing this would require another schema change in testresults.

Ultimately, once we know that the new implementation is correct and that it puts less burden on the database, the old hourly and daily test stats code can be removed. The code could be simplified to work directly per project and with the ignore tasks filter (although this may require changing the index at that point). But I suspect that it would also require some kind of throttling mechanism. 

**This code does not change daily task processing as it is not currently as disruptive.**

jimoleary commented on model/stats/stats_test.go:
```diff
@@ -408,6 +588,8 @@ func (s *statsSuite) TestGenerateDailyTaskStats() {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
+	// qry := bson.M{}
```
dead code

jimoleary commented on model/stats/stats_test.go:
```diff
@@ -877,10 +1072,23 @@ func (s *statsSuite) countDailyTaskDocs() int {
+		end = start.Add(24 * time.Hour)
+	}
+
+	// func (s *statsSuite) getLastTestResult(testStatsID DbTestStatsId) (*testresult.TestResult, error) {
```
dead code.

jimoleary commented on model/stats/stats_test.go:
```diff
@@ -931,10 +1139,67 @@ func (s *statsSuite) getLastHourlyTestStat(testStatsID DbTestStatsId) (*dbTestSt
+
+func (s *statsSuite) validateDbTestStats(testStatsID DbTestStatsId, date time.Time, numPass int, numFail int, avgDurationPass float64, lastUpdate time.Time, hourly bool) {
+	require := s.Require()
+	// var doc *DbTestStats
```
dead code.

jimoleary commented on units/cache_historical_test_data.go:
```diff
@@ -33,6 +33,7 @@ type cacheHistoricalTestDataJob struct {
 	ProjectID       string   `bson:"project_id" json:"project_id" yaml:"project_id"`
 	Requesters      []string `bson:"requesters" json:"requesters" yaml:"requesters"`
 	DisableOldTasks bool     `bson:"disable_old_tasks" json:"disable_old_tasks" yaml:"disable_old_tasks"`
+	UseMerge        bool     `bson:"use_merge,omitempty" json:"use_merge,omitempty" yaml:"use_merge,omitempty"`
```
I may be missing a service flag for configuring use merge. It defaults to off for now.

bsamek commented:
@tychoish Just a note that we can't merge this until we update the database, and we need to figure out what our backup story is on 4.2.

jimoleary commented:
Good point and there are also some indexes that need to be determined and created.

It could be merged and kept disabled which could help testing on staging.

jimoleary commented:
test-cloud is not related to this PR.
test-model-stats and test-units are failing because the backing mongodb is 4.0 and therefore doesn't support $merge.

jimoleary commented:
updated the $merge pipeline to $match on display_name and rebased to move the commit before the 2 test commits (github seems to list the commit chronologically anyway).

jimoleary commented on model/stats/db.go:
```diff
@@ -1114,18 +1266,49 @@ func makeSum(condition bson.M) bson.M {
+        return getLatestTestDoc(dailyTestStatsCollection, projectID, requester, tasks)
+}
+
+func getLatestTestDoc(collection string, projectID string, requester string, tasks []string) (*DbTestStats, error) {
```
This needs to be adjusted. it's not quite right (it may be looking up via display_name but the needs to take the execution_display_name into account too).

tychoish commented:
just for the record, I haven't been looking at this given the last comment and the failing tests. Let me know if/when this is ready for another pass.

bsamek commented:
Given that we don't expect to be able to merge this until after a project to upgrade the database, I'm going to close this PR for now.
